File: 2018/Python/J3.py
"""
d1,d2,d3,d4,d5 = [],[],[],[],[] #SOLUTION 1

d = input().split()
for i in range (len(d)):
    d[i] = int(d[i])

for i in range (5):
    d1.append(sum(d[:i]))
    d2.append(sum(d[1:i]))
    d3.append(sum(d[2:i]))
    d4.append(sum(d[3:i]))
    d5.append(sum(d[4:i]))

for i in range (3,-1,-1):
    d5[i] = sum(d[i:4])
    if i<=2:
        d4[i] = sum(d[i:3])
    if i<=1:
        d3[i] = sum(d[i:2])
    d2[0] = d[0]

for a,b,c,d,e in d1,d2,d3,d4,d5:
    print (a,b,c,d,e)
"""
d = input().split()

d1 = int(d[0]) #SOLUTION 2
d2 = int(d[1])
d3 = int(d[2])
d4 = int(d[3])

print("0",d1,(d1+d2),(d1+d2+d3),(d1+d2+d3+d4))
print(d1,"0",d2,(d2+d3),(d2+d3+d4))
print((d1+d2),d2,"0",(d3),(d3+d4))
print((d1+d2+d3),(d2+d3),d3,"0",d4)
print((d1+d2+d3+d4),(d2+d3+d4),(d3+d4),d4,"0")

File: 2019/Java/J3.java
import java.util.*;
public class J3 {
	public static void main(String args[]) {
        Scanner input = new Scanner(System.in);
        String lines = input.nextLine();
        int L = Integer.parseInt(lines);
        System.out.println(L);
        List<String> lst = new ArrayList<String>();
        for (int i = 0; i<L; i++){
            List<String> lst1 = new ArrayList<String>();
            String text = input.nextLine();
            while (text!=""){
                int num = StringUtils.countMatches(text, text.charAt(0));
                if (num>1){
                    for (int x = 1; x<num; x++){
                        if (text.charAt(x) != text.charAt(x-1)){
                            num = x;
                            break;
                        }
                    }
                }
                lst1.add(num + " " + text.charAt(0));
                text = text.substring(num);
            }
        }
    }
}
/*
L = int(input())
lst = []
for counter in range (L):
    lst1 = []
    text = input()
    while text != "":
        num = text.count(text[0])
        if num > 1:
            for counter in range (1,num):
                if text[counter] != text[counter-1]:
                    num = counter
                    break
        lst1.append(str(num) + " " + text[0])
        text = text[num:]
    lst.append(" ".join(lst1))
for counter in range (len(lst)):
    print (lst[counter])
*/
